localstack/utils/net.py

```diff
@@ -193,17 +201,74 @@ def get_free_udp_port(blocklist: List[int] = None) -> int:
 
 
 def get_free_tcp_port(blocklist: List[int] = None) -> int:
+    """
+    Tries to bind a socket to port 0 and returns the port that was assigned by the system. If the port is
+    in the given ``blocklist``, or the port is marked as reserved in ``dynamic_port_range``, the procedure
+    is repeated for up to 50 times.
+
+    :param blocklist: an optional list of ports that are not allowed as random ports
+    :return: a free TCP port
+    """
     blocklist = blocklist or []
-    for i in range(10):
+    for i in range(50):
         tcp = socket.socket(socket.AF_INET, socket.SOCK_STREAM)
         tcp.bind(("", 0))
         addr, port = tcp.getsockname()
         tcp.close()
-        if port not in blocklist:
+        if port not in blocklist and not dynamic_port_range.is_port_reserved(port):
+            try:
+                dynamic_port_range.mark_reserved(port)
+            except ValueError:
+                # depending on the ephemeral port range of the system, the allocated port may be outside what
+                # we defined as dynamic port range
+                pass
             return port
     raise Exception(f"Unable to determine free TCP port with blocklist {blocklist}")
 
 
+def get_free_tcp_port_range(num_ports: int, max_attempts: int = 50) -> "PortRange":
+    """
+    Attempts to get a contiguous range of free ports from the dynamic port range. For instance,
+    ``get_free_tcp_port_range(4)`` may return the following result: ``PortRange(44000:44004)``.
+
+    :param num_ports: the number of ports in the range
+    :param max_attempts: the number of times to retry if a contiguous range was not found
+    :return: a port range of free TCP ports
+    :raises PortNotAvailableException: if max_attempts was reached to re-try
+    """
+    if num_ports < 2:
+        raise ValueError(f"invalid number of ports {num_ports}")
+
+    def _is_port_range_free(_range: PortRange):
+        for _port in _range:
+            if dynamic_port_range.is_port_reserved(_port) or not port_can_be_bound(_port):
+                return False
+        return True
+
+    attempts = 0
+    while attempts <= max_attempts:
+        attempts += 1
+
+        # try to find a suitable starting point (leave enough space at the end)
+        port_range_start = random.randint(
+            dynamic_port_range.start, dynamic_port_range.end - num_ports - 1
+        )
+        port_range = PortRange(port_range_start, port_range_start + num_ports - 1)
```

Comment:
I'm not sure starting at a random port and checking the range is a great approach. It does not take that long on my machine to scan the entire dynamic port range specified (~330ms) so could we check sequentially, perhaps with some intelligence that if a blocked port is found that we restart the scan from the port above that one?

I'm probably overthinking it though, the chances of there not being a valid range on the first few tries is pretty low.

Reply:
this was the way i had it implemented initially, however went back to the completely random approach.
the problem i ran into was:
* either you start at `32768` every time and scan sequentially, then a subsequent call to `get_free_tcp_port_range(50)` would immediately fail (because max_attempts=50 and the first 50 calls would fail because you previously allocated from that range)
* or you start at a random port in which case you need to check  the boundary condition if you're at the end of the dynamic port range, in which case you need to select a new starting point and you might pick another random port again anyway.

you could implement some sort of exponential forward expansion when a port failed, but now it's getting really involved, and the random port approach should be pretty stable.